### PR TITLE
[chore] Refactor all classes to hooks

### PR DESCRIPTION
**Pull request by dej611**
#### PR Checklist

- [x] My branch is up-to-date with the Upstream `master` branch.
- [x] The unit tests pass locally with my changes (if applicable).
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable).
- [x] I have added necessary documentation (if appropriate).

#### Short description of what this resolves:
Complete the code refactoring from class to hooks.
* [x] add unit tests for classes rewritten with hooks

Some extras:
* Created a parallel hook `useFontLoader` to use as replacement of the `FontLoader` component.
  * Update documentation
* Rewrote some utility functions from `reactstrap` as internal functions
  * [x] Add some unit tests
* Rewrote some constant imports from `reactstrap` as static internal values
* Attempt to remove the SSR check in favour of conditional check (?!)
* Removed unused and undocumented `PasswordInput` 🔥 
  * [x] Add doc for migration just in case somebody found it and was using it

### CODE REVIEW

**Comment by dej611** [comment]
Just for SEO context I'll address here how I made it pass the treeshake check with `agadoo`:
* Moved the remaining React components implemented with `class` with their hook version (see also https://github.com/Rich-Harris/agadoo/issues/12 )
* Re-implemented some initial code used for SSR environment check in alternative way (like use a conditional check for its client-side method)
* Re-written async/await code into thenables (this removed the Babel polyfill about iterators)

File: src/Accordion/AccordionBody.tsx [hunk 1]
```diff
@@ -1,35 +1,51 @@
...
-
 export type AccordionBodyProps = Partial<TransitionProps> & {
   tag?: ElementType;
   className?: string;
   active?: boolean;
   onToggle?: () => void;
 };
 
+// Taken from Reactstrap utils.js which are tken from bootstrap sass
+export const TransitionTimeouts = {
```
Comment (sanyamdogra) [inline src/Accordion/AccordionBody.tsx:16]: Shouldn't we move these to some shared const files and then consume these? These appear to be generic

File: src/Accordion/AccordionBody.tsx [hunk 2]
```diff
@@ -1,35 +1,51 @@
...
+  // $transition-fade
+  Collapse: 350,
+  // $transition-collapse
+  Modal: 300,
+  // $modal-transition
+  Carousel: 600 // $carousel-transition
+};
+
+// taken from Reactstrap utils.js
+export const TransitionsKeys = [
```
Comment (sanyamdogra) [inline src/Accordion/AccordionBody.tsx:27]: Same as above^

**Comment by dej611** [comment]
@baldoalessandro this PR is passing the treeshake check but the bash script is not catching that up correctly. Can you have a look at that?

**Comment by baldoalessandro** [comment]
@dej611 Sure thing, if I find something should I add commits here or open a new PR?

**Comment by dej611** [comment]
> @dej611 Sure thing, if I find something should I add commits here or open a new PR?
> ### Update
> 
> I've found the issue and made a PR to test it in my fork, you can see it here [baldoalessandro#5](https://github.com/baldoalessandro/design-react-kit/pull/5) Let me know if I should open a PR against master or this branch, sorry for the inconvenience 😊

You can create a PR here that targets this PR/branch.

**Comment by dej611** [comment]
@baldoalessandro I've borrowed your fix for #830 as I was already fixing some stuff in the area.
If you prefer to submit your own PR I'll revert the changes and let you push your own PR.

**Comment by baldoalessandro** [comment]
@dej611 no problem :+1:

> EDIT:
>I've tested your fix and it is still buggy. It returns that the PR is tree-shakeable while it's not.
Do you mean that the check in the PR passed in both cases or there was a problem with the exit-code parsing?
In the first case I didn't know you wanted to fail the check but I've seen you already added:

```yaml
      - name: Fail if cannot tree-shake
        if: ${{ steps.agadoo.outputs.status == 'not tree-shakeable' }}
        run: exit 1
```
in case there are still errors I'm happy to help you!

**Comment by dej611** [comment]
> Do you mean that the check in the PR passed in both cases or there was a problem with the exit-code parsing?
> In the first case I didn't know you wanted to fail the check but I've seen you already added:

It was reporting success all the times, as RET was not correctly capturing the correct output from agadoo. So I rewrote the script a bit to read the last line of the output

**Pull request merged**